examples/99_in_tool_selection/ex_in_tool_selection/main.rb

```diff
@@ -0,0 +1,364 @@
+
+      def suspend(view)
+        view.invalidate
+      end
+
+      def onCancel(reason, view)
+        reset_tool
+        view.invalidate
+      end
+
+      def onKeyDown(key, _repeat, _flags, view)
+        @inference_lock_helper.on_keydown(key, view, @mouse_ip, @picked_first_ip)
+        pick_mouse_position(view)
+      end
+
+      def onKeyUp(key, _repeat, _flags, view)
+        @inference_lock_helper.on_keyup(key, view)
+        pick_mouse_position(view)
+      end
+
+      def onMouseMove(flags, x, y, view)
+        if @selection.empty?
+          # If the tool selection is empty, try picking the hovered object.
+          try_select_entity(view, x, y)
+        end
+
+        @mouse_position = Geom::Point3d.new(x, y, 0)
+        pick_mouse_position(view)
+      end
+
+      def onLButtonDown(flags, x, y, view)
+        @mouse_down = Geom::Point3d.new(x, y)
+
+        # If tool selection is empty, pick our hovered entity.
+        if @selection.empty?
+          @selection = view.model.selection.to_a
+        end
+
+        return if @selection.empty?
+
+        # For a Move or Rotate like tool, we can select a starting point in the
+        # same mouse click as we select an entity. For Scale or other tools
+        # where you need to select a handle, this logic is typically slightly
+        # different, and separate click is needed.
+
+        if !picked_first_point?
+          start_move(view)
+        else
+          end_move(view)
+        end
+
+        update_ui
+        view.invalidate
+      end
+
+      def onLButtonUp(flags, x, y, view)
+        if @mouse_down.distance([x, y]) > DRAG_THRESHOLD
+          end_move(view)
+        end
+      end
+
+      # Here we have hard coded a special ID for the move cursor in SketchUp.
+      # Normally you would use `UI.create_cursor(cursor_path, 0, 0)` instead
+      # with your own custom cursor bitmap:
+      #
+      #   CURSOR = UI.create_cursor(cursor_path, 0, 0)
+      MOVE_CURSOR = 641
+      INVALID_CURSOR = 663
+
+      def onSetCursor
+        # Note that `onSetCursor` is called frequently so you should not do much
+        # work here. At most you switch between different cursor representing
+        # the state of the tool.
+
+        # Cursor depends on the tool state and hovered entity.
+        # If the user is at the selection stage and hover a non-selectable
+        # entity, they se an invalid-cursor.
+        # Hovering empty space in SketchUp doesn't show the invalid-cursor,
+        # even if empty space can't be selected.
+        if !picked_first_point? && @hovered_entity && !@valid_hovered_entity
+          UI.set_cursor(INVALID_CURSOR)
+        else
+          UI.set_cursor(MOVE_CURSOR)
+        end
+      end
+
+      def onUserText(text, view)
+        begin
+          distance = text.to_l
+        rescue ArgumentError
+          UI.messagebox('Invalid length')
+          return
+        end
+        direction = picked_points[1] - picked_points[0]
```

Comment:
I would prefer to see `.vector_to` here instead of the minus operator as it much more explicit. Various 3d libraries have various meaning for the math operators on 3d geometry objects - which can be a source of confusion.

Reply:
I'm more used to this syntax but I don't know if one is better than the other. My brain likes when there are fewer tokens to parse.

Reply:
`.vector_to` makes it obvious that you get a vector in return. `-` is not so clear in it's intent. Being unfamiliar with the API or being used to a different 3D API this could be read as subtracting the members and returning a `Point3d`.

`.vector_to` is unambiguous to what it does which is a benefit when used with examples/tutorials. If this had not been an example or tutorial I wouldn't objected to either variant.

Reply:
Changing to vector_to